### ISSUE TITLE
Best way to store pre-computed embeddings?

### ISSUE THREAD

**Issue by camlaedtke** [lucidrains/imagen-pytorch#137]
Hello!

I have been exploring the idea of storing the pre-computed text embeddings of the image captions since it would free up some VRAM during training that is usually taken by the text embedding model. However, the storage requirements seem to be massive - I computed about 10,000 text embeddings using T5 XL and stored each in an individual pickle file, and this took up around 21 GB of storage. Since I am working with millions of image captions, my current method of storing the embeddings is unworkable.

So I am curious, do you know if there is a better way to store the embeddings?

**Comment by lucidrains** [comment]
@camlaedtke highly recommend https://github.com/hristo-vrigazov/mmap.ninja

i may use it, in conjunction with webdatasets, once i circle back to imagen to build out video generation

but as for today, it is back to meditating on attention

**Comment by lucidrains** [comment]
@camlaedtke one of the authors of Laion also have an open sourced embedding library https://github.com/rom1504/embedding-reader

**Comment by lucidrains** [comment]
@camlaedtke great! if you come up with some good code, do consider submitting a PR and being a contributor!

**Comment by camlaedtke** [comment]
@lucidrains Will do! If I come up with anything useful I'd love to contribute. 

I think coming up with an efficient way to read data from an HDD would be especially helpful given the massive amounts of data involved in training these models.

**Comment by lucidrains** [comment]
@camlaedtke do it! participate in this industry-shaking moment https://news.ycombinator.com/item?id=32168190

**Comment by camlaedtke** [comment]
@lucidrains Wow what a fascinating thread. DALL-E and Imagen have really opened up Pandora's box - so many legal, ethical, and industry implications to think through.

**Comment by lucidrains** [comment]
@camlaedtke tip of the iceberg, this will happen to everything soon

video, audio, music, scene representations, holodeck...

**Comment by camlaedtke** [comment]
@lucidrains Indeed. And the consolidation of all these different AI applications into transformer-based models has been fascinating to watch. I can understand why you're such a transformer nerd lol. 

https://twitter.com/karpathy/status/1468370605229547522?s=20&t=zH4GegYp6HJ2Ozu2qfNUdw

**Comment by lucidrains** [comment]
@camlaedtke https://github.com/lucidrains/attention yes, it does feel like we've been handed some cheat code to breaking reality

**Comment by lucidrains** [comment]
better make good use of it with the limited time we have!

**Comment by camlaedtke** [comment]
> @camlaedtke https://github.com/lucidrains/attention yes, it does feel like we've been handed some cheat code to breaking reality

I love this idea you have. We gotta get 3Blue1Brown to make a video about attention, his neural net series was fantastic.

**Comment by camlaedtke** [comment]
> @camlaedtke ok, if you come up with any good data creation scripts, do consider becoming a contributor 😄

Hey, I did some experiments with pre-computed embeddings and thought you might like to see some of my findings. 

The storage requirements aren't as bad as I initially thought, although they are still pretty huge. If I store each T5 XL embedding tensor to an individual pickle file, it will take up around 3TB for the 10.5 million captions in the cc12m dataset. For my purposes, this is fine since it will fit on my external HDD. I know that there are more efficient ways to store the data, but I am not sure if the space savings would be worth the extra work required to integrate it into the input pipeline. Regardless, the amount of storage required for precomputed embeddings appears to be an order of magnitude larger than the original dataset.

If you are interested, here is my quick and dirty code for creating a webdataset that also contains the pickled caption embeddings. This was written specifically for the cc12m dataset (which I downloaded to my SSD), but it will probably work with other webdatasets since they have a similar format. 


```python
def get_emb_tensor(text):
    text_embeds = t5_encode_text([text], name="google/t5-v1_1-xl", return_attn_mask=False)
    return text_embeds.cpu()


def augment_wds(input, output):
    
    # For tracking progress ...
    stats_file = input[:-4] + "_stats.json"
    f = open(stats_file)
    stats = json.load(f)
    f.close()
    count = stats["successes"]
    
    src = wds.DataPipeline(
        wds.SimpleShardList(input),
        wds.tarfile_to_samples(),
        wds.decode("pil"),
        wds.to_tuple("__key__", "jpg;png", "txt", "txt"),
        wds.map_tuple(None, None, None, get_emb_tensor)
    )
    
    with wds.TarWriter(output) as dst:
        for key, img, cap, emb in tqdm(src, total=count, desc=f"Writing {output}"):
            dst.write({"__key__":key, "png":img, "txt":cap, "emb.pyd":emb})
            
input_shards = braceexpand.braceexpand("cc12m/{00000..01242}.tar")
output_shards = braceexpand.braceexpand("file:E:/datasets/cc12m/{00000..01242}.tar")
for input_shard, output_shard in zip(input_shards, output_shards):
    augment_wds(input=input_shard, output=output_shard)

```
The code works, but it is quite slow since it has to compute each embedding tensor one by one. With my current setup (3700X CPU, RTX 3090 GPU, cheap Seagate external HDD), it runs at about 24 iterations per second or around 3-4 days for the full cc12m dataset. I am currently exploring options for speeding up the process - there is a version of the code above that I am working on which computes the embeddings in batches, which should speed things up quite a bit. 

For what it's worth, here is my data input pipeline, which I configured to read directly from my external HDD. 

```python

def pad_tensor(sequences):
    num = len(sequences)
    max_len = max([s.size(1) for s in sequences])
    out_dims = (num, max_len, 2048) 
    out_tensor = sequences[0].data.new(*out_dims).fill_(0)
    for i, tensor in enumerate(sequences):
        length = tensor.size(1)
        out_tensor[i, :length, :] = tensor[:,:,0].permute(1,0)
    return out_tensor


def zero_pad_batch(batch):
    imgs = [item[0] for item in batch]
    embeds = [item[1] for item in batch]
    embeds = pad_tensor(embeds)
    return [imgs, embeds]


preproc = T.Compose([
    T.Resize(256),
    T.RandomHorizontalFlip(),
    T.CenterCrop(256),
    T.ToTensor()
])


cc_dataset = (
    wds.WebDataset("file:E:/datasets/cc12m/{00000..00099}.tar") # adding 'file:' before filepath is necessary on windows
    .shuffle(4096)
    .decode("pilrgb")
    .rename(image="png", embedding="emb.pyd")
    .map_dict(image=preproc)
    .to_tuple("image", "embedding")
)

cc_dataloader = DataLoader(
    dataset = cc_dataset, 
    batch_size = 256, 
    drop_last = True,
    num_workers = 1, # more than 1 worker slows things down from my experiments
    prefetch_factor = 4,
    pin_memory = True,
    collate_fn = zero_pad_batch
)

```

And here is my training loop. I profiled each component of the training loop and logged the results in wandb, and used the "dead_time" metric to see if I was I/O bottlenecked. 

```python
def train(cfg, dataloader, trainer, epoch, i, device):
    loss_arr = []
    n_batches = (cfg["dataset"]["num_images"] // cfg["train"]["batch_size"])
    step_start = time()
    for step, batch in enumerate(dataloader): 
        curr_step = int(n_batches*(epoch-1) + step)

        fetch_start = time()
        images, text_embeds = batch
        images = torch.stack(images, dim=0)
        images = images.to(device)
        fetch_end = time()
        fetch_time = fetch_end-fetch_start
       
        embed_start = time()
        text_embeds = text_embeds.to(device)
        embed_end = time()
        embed_time = embed_end-embed_start
       
        loss_start = time()
        loss = trainer(
            images, 
            text_embeds = text_embeds, 
            unet_number = i, 
            max_batch_size = cfg["train"]["base_unet_max_batch_size"] if i == 1 else cfg["train"]["sr_unet1_max_batch_size"]
        )
        loss_arr.append(loss)
        loss_end = time()
        loss_time = loss_end-loss_start
        
        
        update_start = time()
        trainer.update(unet_number = i)
        update_end = time()
        update_time = update_end-update_start
        
        step_end = time()
        step_time = step_end-step_start
        
        dead_time = step_time - fetch_time - embed_time - loss_time - update_time


        if step % cfg["train"]["checkpoint_rate"] == 0 and step !=0 and not math.isnan(loss): 
            log_sample_images(cfg, trainer, i)
            trainer.save(cfg["train"]["checkpoint_path"])
            print()
            
        wandb.log({
            f"Train {i} Step": curr_step,
            f"Train Loss {i}": loss, 
            f"Step time {i}": step_time, 
            f"Img load time {i}": fetch_time, 
            f"Embed load time {i}": embed_time,
            f"Loss time {i}": loss_time,
            f"Update time {i}": update_time,
            f"Dead time {i}": dead_time
        })
        print(f"\rTrain Step {step+1}/{n_batches} --- Loss: {loss:.4f} ", end='')
        step_start = time()

    return trainer, loss_arr

```

I ran the training loop for around 900 steps, and you can see it logged here: https://wandb.ai/camlaedtke/imagen/runs/39sp4207?workspace=user-camlaedtke

Surprisingly, I was not I/O bottlenecked for most of the training even when using a single thread, with a shuffle buffer size of 4096 and batch size of 256. And I was also able to free up around 6 GB of VRAM because I didn't have to load the text encoder model. However, I still need to figure out why the dataloader periodically slows down, and explore ways to further optimize things. 

Anyways, I hope some of this is useful to you. I'd love to hear your thoughts!
